[export][function schema] support exporting hop with function schema argument

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #151067
* #152248
* #152247
* #152246
* #152245
* #152244
* __->__ #152073
* #152072

We need to make function schema proxyable to trace a the auto_functionalized hop that takes function schema as inputs.  The implementation basically follows how we support torchbind object:

1. upon seeing an untracked function schema arg, we creates a constant get_attr node
2. we track the function schema argument in export to support lift/unlift.
3. we need to support serde for functional schema. We'll add support for this in follow-up PRs.

However, compared with torchbind object:
1. we don't need a dynamo implementation, because the function schema is added when we auto_functionalize a hop to the argument of auto_functionalized. One potential use case is users re-traces an exported program with strict mode. Since non-strict is the default now, we don't see a use case yet.
2. we don't need an inductor implementation, because the function schema will go away after auto_functionalized re-inplacing pass.

edit: we greatly simplifies (and generalizes) the implementation following @zou3519 's suggestion of using pytree.register_constant


cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

ydwu4 committed:
[export][function schema] support exporting hop with function schema argument

[ghstack-poisoned]

ydwu4 committed:
Update on "[export][function schema] support exporting hop with function schema argument"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

ydwu4 committed:
Update on "[export][function schema] support exporting hop with function schema argument"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

ydwu4 committed:
Update on "[export][function schema] support exporting hop with function schema argument"


We need to make function schema proxyable to trace a the auto_functionalized hop that takes function schema as inputs.  The implementation basically follows how we support torchbind object:

1. upon seeing an untracked function schema arg, we creates a constant get_attr node
2. we track the function schema argument in export to support lift/unlift.
3. we need to support serde for functional schema. We'll add support for this in follow-up PRs.

However, compared with torchbind object:
1. we don't need a dynamo implementation, because the function schema is added when we auto_functionalize a hop to the argument of auto_functionalized. One potential use case is users re-traces an exported program with strict mode. Since non-strict is the default now, we don't see a use case yet.
2. we don't need an inductor implementation, because the function schema will go away after auto_functionalized re-inplacing pass.


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv 

[ghstack-poisoned]

ydwu4 committed:
Update on "[export][function schema] support exporting hop with function schema argument"


We need to make function schema proxyable to trace a the auto_functionalized hop that takes function schema as inputs.  The implementation basically follows how we support torchbind object:

1. upon seeing an untracked function schema arg, we creates a constant get_attr node
2. we track the function schema argument in export to support lift/unlift.
3. we need to support serde for functional schema. We'll add support for this in follow-up PRs.

However, compared with torchbind object:
1. we don't need a dynamo implementation, because the function schema is added when we auto_functionalize a hop to the argument of auto_functionalized. One potential use case is users re-traces an exported program with strict mode. Since non-strict is the default now, we don't see a use case yet.
2. we don't need an inductor implementation, because the function schema will go away after auto_functionalized re-inplacing pass.


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv 

[ghstack-poisoned]

ydwu4 committed:
Update on "[export][function schema] support exporting hop with function schema argument"


We need to make function schema proxyable to trace a the auto_functionalized hop that takes function schema as inputs.  The implementation basically follows how we support torchbind object:

1. upon seeing an untracked function schema arg, we creates a constant get_attr node
2. we track the function schema argument in export to support lift/unlift.
3. we need to support serde for functional schema. We'll add support for this in follow-up PRs.

However, compared with torchbind object:
1. we don't need a dynamo implementation, because the function schema is added when we auto_functionalize a hop to the argument of auto_functionalized. One potential use case is users re-traces an exported program with strict mode. Since non-strict is the default now, we don't see a use case yet.
2. we don't need an inductor implementation, because the function schema will go away after auto_functionalized re-inplacing pass.


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv 

[ghstack-poisoned]

ydwu4 commented on torch/utils/_pytree.py:
```diff
@@ -1211,6 +1211,11 @@ def unflatten(self, leaves: Iterable[Any]) -> PyTree:
 
         return unflatten_fn(child_pytrees, self.context)
 
+    def __hash__(self) -> int:
```
We cannot use the default hashing since list is not hashable. We need to hash the TreeSpec so that it could be put in to tensor_attrs map in _symbolic_trace.

XuehaiPan replied:
Could we implement this by dispatching the `self.type`?

For example, the `dict` tree spec is actually order-sensitive. It can be hashed via:

```python
if self.type is dict:
    return hash((self.type, tuple(self.context), tuple(self.children_specs)))
```

Same argument for `tuple`, `namedtuple`, `list`, `OrderedDict`, `defaultdict`, `deque`, and `torch.return_types.*`.

ydwu4 replied:
hm...I feel this change is beyond the scope of this pr and probably need some more discussion and testing. Mind we add it as a TODO and implement it in a sperate PR?

ydwu4 replied:
Upon a second thought, is it that you want to support the hashing of the listed types with `hash((self.type, tuple(self.context), tuple(self.children_specs)))` and raise notimplemented for other cases?

ydwu4 committed:
Update on "[export][function schema] support exporting hop with function schema argument"


We need to make function schema proxyable to trace a the auto_functionalized hop that takes function schema as inputs.  The implementation basically follows how we support torchbind object:

1. upon seeing an untracked function schema arg, we creates a constant get_attr node
2. we track the function schema argument in export to support lift/unlift.
3. we need to support serde for functional schema. We'll add support for this in follow-up PRs.

However, compared with torchbind object:
1. we don't need a dynamo implementation, because the function schema is added when we auto_functionalize a hop to the argument of auto_functionalized. One potential use case is users re-traces an exported program with strict mode. Since non-strict is the default now, we don't see a use case yet.
2. we don't need an inductor implementation, because the function schema will go away after auto_functionalized re-inplacing pass.

edit: we greatly simplifies (and generalizes) the implementation following zou3519 's suggestion of using pytree.register_constant


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv 

[ghstack-poisoned]

ydwu4 committed:
Update on "[export][function schema] support exporting hop with function schema argument"


We need to make function schema proxyable to trace a the auto_functionalized hop that takes function schema as inputs.  The implementation basically follows how we support torchbind object:

1. upon seeing an untracked function schema arg, we creates a constant get_attr node
2. we track the function schema argument in export to support lift/unlift.
3. we need to support serde for functional schema. We'll add support for this in follow-up PRs.

However, compared with torchbind object:
1. we don't need a dynamo implementation, because the function schema is added when we auto_functionalize a hop to the argument of auto_functionalized. One potential use case is users re-traces an exported program with strict mode. Since non-strict is the default now, we don't see a use case yet.
2. we don't need an inductor implementation, because the function schema will go away after auto_functionalized re-inplacing pass.

edit: we greatly simplifies (and generalizes) the implementation following zou3519 's suggestion of using pytree.register_constant


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv 

[ghstack-poisoned]

XuehaiPan commented on torch/fx/_symbolic_trace.py:
```diff
@@ -395,17 +400,22 @@ def create_arg(self, a: Any) -> "Argument":
+                    base_name = "_tensor_constant"
+                elif isinstance(a, (FakeScriptObject, ScriptObject)):
+                    base_name = "_torchbind_obj"
+                elif isinstance(a, torch.utils._pytree.TreeSpec):
```
```suggestion
                elif isinstance(a, pytree.TreeSpec):
```

ydwu4 committed:
Update on "[export][function schema] support exporting hop with function schema argument"


We need to make function schema proxyable to trace a the auto_functionalized hop that takes function schema as inputs.  The implementation basically follows how we support torchbind object:

1. upon seeing an untracked function schema arg, we creates a constant get_attr node
2. we track the function schema argument in export to support lift/unlift.
3. we need to support serde for functional schema. We'll add support for this in follow-up PRs.

However, compared with torchbind object:
1. we don't need a dynamo implementation, because the function schema is added when we auto_functionalize a hop to the argument of auto_functionalized. One potential use case is users re-traces an exported program with strict mode. Since non-strict is the default now, we don't see a use case yet.
2. we don't need an inductor implementation, because the function schema will go away after auto_functionalized re-inplacing pass.

edit: we greatly simplifies (and generalizes) the implementation following zou3519 's suggestion of using pytree.register_constant


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv 

[ghstack-poisoned]

ydwu4 committed:
Update on "[export][function schema] support exporting hop with function schema argument"


We need to make function schema proxyable to trace a the auto_functionalized hop that takes function schema as inputs.  The implementation basically follows how we support torchbind object:

1. upon seeing an untracked function schema arg, we creates a constant get_attr node
2. we track the function schema argument in export to support lift/unlift.
3. we need to support serde for functional schema. We'll add support for this in follow-up PRs.

However, compared with torchbind object:
1. we don't need a dynamo implementation, because the function schema is added when we auto_functionalize a hop to the argument of auto_functionalized. One potential use case is users re-traces an exported program with strict mode. Since non-strict is the default now, we don't see a use case yet.
2. we don't need an inductor implementation, because the function schema will go away after auto_functionalized re-inplacing pass.

edit: we greatly simplifies (and generalizes) the implementation following zou3519 's suggestion of using pytree.register_constant


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv 

[ghstack-poisoned]

XuehaiPan commented on torch/utils/_pytree.py:
```diff
@@ -1212,9 +1212,10 @@ def unflatten(self, leaves: Iterable[Any]) -> PyTree:
-        if not isinstance(self.context, ConstantNode):
-            raise NotImplementedError("Cannot hash non-ConstantNode TreeSpec.")
-        return hash((self.type, self.context))
+        if isinstance(self.context, ConstantNode):
+            return hash((self.type, self.context))
+        else:
+            return hash((self.type, tuple(self.context), tuple(self.children_specs)))
```
```suggestion
        node_type = self.type
        if node_type is defaultdict:
            default_factory, dict_context = self.context
            hashable_context = (default_factory, tuple(dict_context))
        elif node_type in (dict, OrderedDict):
            hashable_context = tuple(self.context)
        elif node_type in BUILTIN_TYPES:
            hashable_context = self.context
        elif isinstance(self.context, ConstantNode):
            hashable_context = self.context.value
        else:
            # The context for user-defined node types might not be hashable.
            # Ignore it for hashing.
            # This does not break the correctness that equal objects implies
            # same hash. This might increase the hash collision rate, but we
            # don't care about that.
            hashable_context = None
        return hash((node_type, hashable_context, tuple(self.children_specs)))
```

ydwu4 replied:
sure, ok with not having a test lol?

ydwu4 committed:
Update on "[export][function schema] support exporting hop with function schema argument"


We need to make function schema proxyable to trace a the auto_functionalized hop that takes function schema as inputs.  The implementation basically follows how we support torchbind object:

1. upon seeing an untracked function schema arg, we creates a constant get_attr node
2. we track the function schema argument in export to support lift/unlift.
3. we need to support serde for functional schema. We'll add support for this in follow-up PRs.

However, compared with torchbind object:
1. we don't need a dynamo implementation, because the function schema is added when we auto_functionalize a hop to the argument of auto_functionalized. One potential use case is users re-traces an exported program with strict mode. Since non-strict is the default now, we don't see a use case yet.
2. we don't need an inductor implementation, because the function schema will go away after auto_functionalized re-inplacing pass.

edit: we greatly simplifies (and generalizes) the implementation following zou3519 's suggestion of using pytree.register_constant


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv 

[ghstack-poisoned]

ydwu4 committed:
Update on "[export][function schema] support exporting hop with function schema argument"


We need to make function schema proxyable to trace a the auto_functionalized hop that takes function schema as inputs.  The implementation basically follows how we support torchbind object:

1. upon seeing an untracked function schema arg, we creates a constant get_attr node
2. we track the function schema argument in export to support lift/unlift.
3. we need to support serde for functional schema. We'll add support for this in follow-up PRs.

However, compared with torchbind object:
1. we don't need a dynamo implementation, because the function schema is added when we auto_functionalize a hop to the argument of auto_functionalized. One potential use case is users re-traces an exported program with strict mode. Since non-strict is the default now, we don't see a use case yet.
2. we don't need an inductor implementation, because the function schema will go away after auto_functionalized re-inplacing pass.

edit: we greatly simplifies (and generalizes) the implementation following zou3519 's suggestion of using pytree.register_constant


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv 

[ghstack-poisoned]

ydwu4 committed:
Update on "[export][function schema] support exporting hop with function schema argument"


We need to make function schema proxyable to trace a the auto_functionalized hop that takes function schema as inputs.  The implementation basically follows how we support torchbind object:

1. upon seeing an untracked function schema arg, we creates a constant get_attr node
2. we track the function schema argument in export to support lift/unlift.
3. we need to support serde for functional schema. We'll add support for this in follow-up PRs.

However, compared with torchbind object:
1. we don't need a dynamo implementation, because the function schema is added when we auto_functionalize a hop to the argument of auto_functionalized. One potential use case is users re-traces an exported program with strict mode. Since non-strict is the default now, we don't see a use case yet.
2. we don't need an inductor implementation, because the function schema will go away after auto_functionalized re-inplacing pass.

edit: we greatly simplifies (and generalizes) the implementation following zou3519 's suggestion of using pytree.register_constant


cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv 

[ghstack-poisoned]
